cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -428,190 +425,128 @@ func getAPIGroupPreferredVersion(client discovery.DiscoveryInterface, APIGroup s
-	return providerIDs, nil
-}
-
-func (c *machineController) filterAllMachineSets(f machineSetFilterFunc) error {
-	return c.filterMachineSets(metav1.NamespaceAll, f)
-}
-
-func (c *machineController) filterMachineSets(namespace string, f machineSetFilterFunc) error {
-	machineSets, err := c.listMachineSets(namespace, labels.Everything())
-	if err != nil {
-		return nil
-	}
-	for _, machineSet := range machineSets {
-		if err := f(machineSet); err != nil {
-			return err
+		nodeRefName, found, err := unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "name")
+		if err != nil {
+			return nil, err
 		}
-	}
-	return nil
-}
 
-func (c *machineController) machineSetNodeGroups() ([]*nodegroup, error) {
-	var nodegroups []*nodegroup
+		if found {
+			node, err := c.findNodeByNodeName(nodeRefName)
+			if err != nil {
+				return nil, fmt.Errorf("unknown node %q", nodeRefName)
+			}
 
-	if err := c.filterAllMachineSets(func(machineSet *MachineSet) error {
-		if machineSetHasMachineDeploymentOwnerRef(machineSet) {
-			return nil
-		}
-		ng, err := newNodegroupFromMachineSet(c, machineSet)
-		if err != nil {
-			return err
-		}
-		if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
-			nodegroups = append(nodegroups, ng)
+			if node != nil {
+				providerIDs = append(providerIDs, node.Spec.ProviderID)
+			}
 		}
-		return nil
-	}); err != nil {
-		return nil, err
 	}
 
-	return nodegroups, nil
+	klog.V(4).Infof("nodegroup %s has nodes %v", scalableResource.GetName(), providerIDs)
+
+	return providerIDs, nil
 }
 
-func (c *machineController) machineDeploymentNodeGroups() ([]*nodegroup, error) {
-	machineDeployments, err := c.listMachineDeployments(metav1.NamespaceAll, labels.Everything())
+func (c *machineController) nodeGroups() ([]*nodegroup, error) {
+	scalableResources, err := c.listScalableResources()
 	if err != nil {
 		return nil, err
 	}
 
-	var nodegroups []*nodegroup
+	nodegroups := make([]*nodegroup, 0, len(scalableResources))
 
-	for _, md := range machineDeployments {
-		ng, err := newNodegroupFromMachineDeployment(c, md)
+	for _, r := range scalableResources {
+		ng, err := newNodeGroupFromScalableResource(c, r)
 		if err != nil {
 			return nil, err
 		}
-		// add nodegroup iff it has the capacity to scale
-		if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(md.Spec.Replicas, 0) > 0 {
+
+		if ng != nil {
 			nodegroups = append(nodegroups, ng)
 		}
 	}
-
 	return nodegroups, nil
 }
 
-func (c *machineController) nodeGroups() ([]*nodegroup, error) {
-	machineSets, err := c.machineSetNodeGroups()
-	if err != nil {
-		return nil, err
-	}
-
-	if c.machineDeploymentResource != nil {
-		machineDeployments, err := c.machineDeploymentNodeGroups()
-		if err != nil {
-			return nil, err
-		}
-		machineSets = append(machineSets, machineDeployments...)
-	}
-
-	return machineSets, nil
-}
-
 func (c *machineController) nodeGroupForNode(node *corev1.Node) (*nodegroup, error) {
```

Comment:
Need to add test cases for this method to verify that autodiscovery config is doing the right thing.